Add a template parameter to override auto_create_index value

pugnascotia opened the pull request:
Closes #20640.

This PR introduces a new parameter to v2 templates, `allow_auto_create`, which allows templates to override the cluster setting `auto_create_index`. Notes:

   * `AutoCreateIndex` now looks for a matching v2 template, and if its `allow_auto_create` setting is true, it overrides the usual logic.
   * `TransportBulkAction` previously used `AutoCreateIndex` to check whether missing indices should be created. After talking to @martijnvg, we now rely on `AutoCreateAction`, which was already differentiating between creating indices and creating data streams. I've updated `AutoCreateAction` to use `AutoCreateIndex`.
   * Most of the Java file changes are due to introducing an extra constructor parameter to `ComposableIndexTemplate`.
   * I've added the new setting to various x-pack templates
   * I added a YAML test to check that watches can be created even when `auto_create_index` is `false`.

pugnascotia committed:
Override auto-create index behaviour via a template flag

pugnascotia committed:
WIP - trying to test watcher

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Fixes

pugnascotia committed:
Remove logging

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia commented:
@martijnvg draft PR as discussed in Slack.

cc @jaymode for your interest.

martijnvg commented on server/src/main/java/org/elasticsearch/action/support/AutoCreateIndex.java:
```diff
@@ -67,6 +62,13 @@ public boolean shouldAutoCreate(String index, ClusterState state) {
+
+        // Templates can override the AUTO_CREATE_INDEX_SETTING setting
+        final ComposableIndexTemplate template = findTemplate(index, state.metadata());
+        if (template != null && template.getAllowAutoCreate()) {
```
We talked about the auto create index logic and the allow auto create template flag yesterday. This logic is invoked from the bulk action and is evaluated on the coordinating node. While the decision the auto create an index or data stream is done in the `AutoCreateAction` on the elected master node. 

The reason why we decided to make the decision to auto create an index or data stream in the `AutoCreateAction` and not here is, because if for some reason a node is lacking behind and doesn't have all the templates then an index is auto created instead of a data stream. Fixing the erroneous situation requires a lot of manual work, hence we moved making this decision to elected master node.

The question is whether the auto create index check (with this enhancement) should be moved to `AutoCreateAction` class too, so that all the decisions are made on the elected master node. 

I'm currently thinking the following:
* In the case a coordinating node doesn't have the latest `action.auto_create_index` setting values or composable index templates then the worst that can happen is a new few write requests fail, because the target index doesn't exist.
* Currently if `action.auto_create_index` is set to `false` then also data streams are no longer created. I think this is a bug, because the name of the setting implies auto creation of indices only. 
* If this logic is moved to `AutoCreateAction` class then the overhead is that remote requests are made to the elected master node in the case target indices don't exist. Most requests will index into an existing data stream or index, so the cost is acceptable.
* I remember that we eventually want to deprecate and remove the `action.auto_create_index` setting, but I may be mistaken.

If we are going to remove `action.auto_create_index` with the new allow auto create template flag then I think let's keep it here? And then maybe also add logic here that if a non existing index matches with a template that auto create data streams then checking the `action.auto_create_index` setting should be ignored?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Use AutoCreateIndex in AutoCreateAction

pugnascotia committed:
Imports

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Fixes

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Fix compile error

pugnascotia committed:
Tweak ComposableIndexTemplate serialisation under code changes land in 7.10

pugnascotia committed:
Fix auto create index bug

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Test fix

pugnascotia committed:
Handle null Boolean

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Imports

pugnascotia commented:
@martijnvg can you take a look at what I've done here with `AutoCreateAction`?

martijnvg commented:
@pugnascotia I like the overall direction of this PR. Moving the auto create index check to AutoCreateAction is a good change, since that fixes the problem that data streams are no longer auto created if `action.auto_create_index` is set to `false`, at the cost of an additional hop to elected master node (this cost is neglectable, as we only redirect to auto create action is index is missing and in order to create the missing index the extra hop to elected master node needs to be done anyway).

pugnascotia committed:
Checkstyle

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Post-merge fix

pugnascotia committed:
Update synthentics template to add allow_auto_create

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Indices APIs)

pugnascotia commented:
I'm not sure what I ought to do next here, so...proper review please? 😀

jaymode commented on server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:
```diff
@@ -132,6 +138,11 @@ public ComposableIndexTemplate(StreamInput in) throws IOException {
         } else {
             this.dataStreamTemplate = null;
         }
+        if (in.getVersion().after(Version.V_7_10_0)) {
```
Typically, we set this to V.8_0_0 until after we backport

pugnascotia replied:
Ah, I wasn't sure what the convention was here. Does that mean that the backport has `V_7_10_0` and master has `V_8_0_0`, or do we also adjust master?

jaymode commented on x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalExporter.java:
```diff
@@ -430,7 +430,7 @@ private boolean hasTemplate(final ClusterState clusterState, final String templa
     // FIXME this should use the IndexTemplateMetadataUpgrader
     private void putTemplate(String template, String source, ActionListener<AcknowledgedResponse> listener) {
-        logger.debug("installing template [{}]", template);
+        logger.warn("installing template [{}]", template);
```
we probably don't want this change to go in?

pugnascotia replied:
Indeed 👍

jaymode reviewed (commented):
Looking good. I left a couple of comments and I think we need some documentation updates too.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Address review feedback

martijnvg commented on x-pack/plugin/core/src/main/resources/ilm-history.json:
```diff
@@ -80,6 +80,7 @@
         }
       }
   },
+  "allow_auto_create": true,
```
I think that `allow_auto_create` doesn't need to be defined? Since auto creation is allowed by default?

pugnascotia replied:
This is to handle the situation where `auto_create_index` is set to false (or some restrictive pattern). In that case, having `allow_auto_create` set to true means that ES will still create the index, overriding `auto_create_index`. Is the name of this new field confusing, maybe?

martijnvg replied:
No, the name is not confusing. I forgot, that the new setting can also be used to overwrite anything that is specified in `auto_create_index` setting.

martijnvg reviewed (commented):
Left a question.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Specify the same version for ser/deser

Honestly, what was I thinking.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Fix stupid typo

pugnascotia commented:
CI is green now on this PR - comments welcome.

jaymode reviewed (changes requested):
I left a comment that I think we need to address. Also, do we need to increment the version used by these templates since we're adding the new value to them.

gwbrown commented on x-pack/plugin/watcher/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/watcher/execute_watch/90_auto_create_index.yml:
```diff
@@ -0,0 +1,52 @@
+---
```
I like this test case, but I think it should be moved to the `put_watch` directory, rather than being in `execute_watch`.

AthenaEryma reviewed (commented):
Looks good overall, but I left a few relatively minor comments. I also agree with Jay's last comment.

And there wasn't a great place to put this, but a basic Java REST test would be good I think - the only integration test for this functionality at the moment is tied to Watcher.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Address review comments

pugnascotia commented:
> do we need to increment the version used by these templates since we're adding the new value to them.

My reading of [this comment in StackTemplateRegistry.java](https://github.com/elastic/elasticsearch/blob/d21ddb57a3a4e3b10d230ac152df2cebb852f6f0/x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackTemplateRegistry.java#L24) suggests that we shouldn't, but I may be mistaken?

pugnascotia commented:
Actually that only applies to `xpack.stack.template.version` - other templates could be incremented.

pugnascotia committed:
Increment some template versions due to the new setting

pugnascotia committed:
Add Java REST tests for auto-creating indices

pugnascotia committed:
Checkstyle

gwbrown commented:
Thanks, looks like you've addressed all my comments - just one more thing. Could you add this parameter to the [Component Template docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-component-template.html)?

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Update component template docs

pugnascotia commented:
Thanks @gwbrown, I added some docs. I also merged William's changes around creating system indices, so a check over the merge would be appreciated 🙏

pugnascotia committed:
Checkstyle

pugnascotia commented:
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/packaging-sample-windows

AthenaEryma reviewed (commented):
Left one nitpick and a question - LGTM assuming that you're ahead of me on the interaction with data streams.

pugnascotia committed:
Docs tweak

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia commented on docs/reference/indices/put-component-template.asciidoc:
```diff
@@ -112,6 +112,16 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=settings]
+<<index-creation,`action.auto_create_index`>> cluster setting. If set to
+`true` in a template, then indices can be automatically created using that
+template even if auto-creation of indices is disabled via
+`actions.auto_create_index`.  If set to `false`, then indices matching the
```
Ah thanks, I'm not a double spacer by habit, that must have come from re-wrapping the lines.

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/history/SnapshotLifecycleTemplateRegistry.java:
```diff
@@ -37,8 +37,9 @@
-    // version 4: add `allow_auto_create` setting
-    public static final int INDEX_TEMPLATE_VERSION = 4;
+    // version 4:converted data stream
+    // version 5: add `allow_auto_create` setting
```
According to the [`AutoCreateAction` class](https://github.com/elastic/elasticsearch/blob/41b2fc6588c839ff91af7e354eed44726295f523/server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java#L120), data streams are always auto-created. I'm not sure whether that answers your question?

pugnascotia replied:
This is an interesting point. Right now, there's no benefit to adding `allow_auto_create` to data stream templates. However, it seems to me that the user experience would be more consistent if the auto-creation of data streams was also subject to the `auto_create_index` setting. Then an administrator could stop their users from e.g. creating a slew of `logs-*-*` data streams.

@martijnvg I believe you added the code to auto-create data streams - what is your view here?

pugnascotia replied:
I've changed `AutoCreateAction` so that data streams are now also subject to `auto_create_index`, and added an integration test. I tried reworking the exist auto-create IT, but as it's under `qa` the data stream API wasn't available.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

pugnascotia committed:
Make data stream subject to auto_create_index

pugnascotia committed:
Add tests for auto-creating data streams

pugnascotia committed:
Formatting

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 20640-auto-create-templates

mayya-sharipova committed:
Fix max/min aggs for unsigned_long (#63904)

Max and min aggs were producing wrong results for unsigned_long field
if field was indexed. If field is indexed for max/min aggs instead of
field data, we use values from indexed Points, values of which
are derived using method pointReaderIfPossible. Before
UnsignedLongFieldType#pointReaderIfPossible was incorrectly
producing values, as it failed to shift them back to original
values.

This patch fixes method pointReaderIfPossible to produce
correct original values.

Relates to #60050

gwbrown commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -116,25 +116,25 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                    final String indexName = indexNameExpressionResolver.resolveDateMathExpression(request.index());
+
+                    // This will throw an exception if the index or data stream does not exist and creating it is prohibited.
+                    final boolean shouldAutoCreate = autoCreateIndex.shouldAutoCreate(indexName, currentState);
```
Moving this call here has the side effect of also applying `action.auto_create_index` to data streams if the template does not have `allow_auto_create` set, which I don't think is what we wanted to do.

pugnascotia replied:
I must have misunderstood - to clarify, were you expecting that all I was going to do was apply the value of `allow_auto_create` to data streams, as opposed to also making them subject to `auto_create_index`? I'm happy to implement either, it just seemed strange that a user could control auto-creation for indices, but not for data streams. @jaymode what's your feeling here?

gwbrown replied:
At this point, making `action.auto_create_index` apply to data streams would be a breaking change, no? With `allow_auto_create`, that's a new feature so we can define the behavior however we want, but we have to be more careful with behavior we've already shipped.

pugnascotia replied:
Ah, that's a good point. I'll rework the code.

pugnascotia replied:
I've reworked `AutoCreateAction` - what do you think about the docs for `allow_auto_create`, is it worth discussing the difference between indices and data streams w.r.t the new template setting?

martijnvg replied:
I think the fact that `action.auto_create_index` controls whether data streams can be auto created can be seen as a bug. Something that we missed during development. I would prefer that `action.auto_create_index` doesn't affect the auto creation of data streams.

martijnvg replied:
Never mind my comment. The PR is doing what I suggested in my comment.

williamrandolph committed:
Add APM configuration index to Kibana system indices (#63756)

* Add APM index to Kibana system indices, making it 
accessible through the _kibana endpoint and giving it the
same access privileges as the other Kibana system indices.
* Parameterize kibana system index tests by index name

dimitris-athanasiou committed:
[ML] Rename evaluation metric result fields to value (#63809)

Renames data frame analytics _evaluate API results as follows:

  - per class accuracy renamed from `accuracy` to `value`
  - per class precision renamed from `precision` to `value`
  - per class recall renamed from `recall` to `value`
  - auc_roc `score` renamed to `value` for both outlier detection and classification

breskeby committed:
Update gradle wrapper to Gradle 6.7 (#62386)

- Ignore Memory Manager thread in Gradle thread filter
- Update gradle wrapper to 6.7 GA release

tlrx committed:
Add snapshot shard size based test in DiskThresholdDeciderTests (#63546)

This commit adds a test in DiskThresholdDeciderTests that verifies
 the allocation of a snapshot recovery source based shard in the 
situation where the snapshot shard size was successfully provided 
by the SnapshotInfoService introduced in #61906 and when the 
service failed to provide the size.

Relates #61906

javanna committed:
DocumentMapperParser to no longer depend directly on MapperService (#63850)

This change was mainly triggered by the need for `MapperService` to pass `this` during its constructor when creating `DocumentMapperParser`. Also, `MapperService` is carried around in some places where only a subset of it is needed.

With this change we rather carry around the components that are strictly needed, in a couple of cases functions that `MapperService` provides, which helps clarifying the dependency between `DocumentMapperParser`, `DocumentMapper` and `MapperService`, as well as removing the need for MapperService to pass `this` to `DocumentMapperParser`

jimczi committed:
Composite aggregation must check live docs when the index is sorted (#63864)

This change ensures that the live docs are checked in the composite aggregator
when the index is sorted.

breskeby committed:
Fix build tools integTest failure after vault update (#63914)

iverase committed:
Upgrade to lucene-8.7.0-snapshot-72d8528c3a6 (#63912)

benwtrent committed:
[ML] fix inference binary classification predication label and feature importance (#63688)

When calculating feature importance, the leaf values directly correlate the value of the importance.

Consequently, positive leaf values -> positive feature importance

negative leaf values -> negative feature importance.

It follows that for binary classification, this is done such that the importance relates to the leaf values, which relate directly to the "probability of class 1".

So, the feature importance calculated is always for the importance as it relates to class 1.

The inverse is the importance as it relates to class 0.

Commit:
Move clone snapshot API page. (#63902)

javanna committed:
Remove two redundant DocumentMapper methods (#63922)

DocumentMapper exposes field types and object mappers through specific getter methods, that call the corresponding getters exposed by MappingLookup. MappingLookup is though exposed directly by DocumentMapper, hence there is no need for additional methods other than the one to retrieve the mapping lookup object.

iverase committed:
make sure AggregationTest creates reduced aggregations. (#63931)

The test calls InternalAggregationTestCase#createTestInstanceForXContent() instead of InternalAggregationTestCase#createTestInstance() (The method needs to become public)

benwtrent committed:
[ML] adding new flag exclude_generated that removes generated fields in GET config APIs (#63899)

When exporting and cloning ml configurations in a cluster it can be
frustrating to remove all the fields that were generated by
the plugin. Especially as the number of these fields change
from version to version.

This flag, exclude_generated, allows the GET config APIs to return
configurations with these generated fields removed.

APIs supporting this flag: 
- GET _ml/anomaly_detection/<job_id>
- GET _ml/datafeeds/<datafeed_id>
- GET _ml/data_frame/analytics/<analytics_id>

The following fields are not returned in the objects:

- any field that is not user settable (e.g. version, create_time)
- any field that is a calculated default value (e.g. datafeed chunking_config)
- any field that is automatically set via another Elastic stack process (e.g. anomaly job custom_settings.created_by)

relates to #63055

benwtrent committed:
[Transform] add new exclude_generated flag to GET transform (#63093)

This adds a new flag `exclude_generated` for GET transform API.

This flag is useful for when a transform needs to be cloned within a cluster or exported/imported between clusters.

It removes certain fields that are not able to be set via the PUT api (e.g. version, create_time).

relates https://github.com/elastic/elasticsearch/issues/63055

jakelandis committed:
Gradle - compatible REST test plugin - adopt bwc artifact (#63629)

Use newly exposed bwc:minor / checkout artifact as opposed to coding directly against
the output of a bwc:minor checkout task. The copy test/api tasks have been updated
to lazily compute the subset of the files from the configuration. This is to facilitate
compatible testing which needs to filter singular bwc:minor checkout in core/x-pack/project
set of files. By default (unless explicitly requested) the copy test/api tasks will continue
to use the full fileTree of the configuration.

related: #63173

javanna committed:
Remove documentMapperParser method from MapperService (#63938)

MapperService allows to retrieve its internal DocumentMapperParser instance. Such method is only used in tests, and always to parse mappings which is already exposed by MapperService through a specific parse method.

This commit removes the getter for DocumentMapperParser from MapperService in favour of calling MapperService#parse

javanna committed:
Minor FieldTypeLookup tweaks (#63944)

Remove the default constructor, make the class final and adjust visibility of its methods: if the class is package private, it makes little sense to have public methods.

Commit:
[Transform] add support for unsigned_long data type (#63940)

add support for unsigned_long, which required a change in
writing out integer results properly, because coerce is not
supported for unsigned_long

fixes #63871

nik9000 committed:
Make agg test less confusing (#63952)

We have an `@After` annotated method in `AggregatorTestCase` that cleans
up releasibles. But it was `private`! Confusingly, it seemed to be
working! I'm not sure why. This makes it public which is a little more
sensible.

nik9000 committed:
Add a test for regex usage to runtime fields (#63951)

Now that we've got regexes enabled by default (#63029) this adds a test
to runtime fields just to make sure that it works with regexes. It does,
but this adds a test to make sure it continues to work.

dimitris-athanasiou committed:
[ML] Extend default evaluation metrics to all available (#63939)

This commit extends the set of default metrics for the
data frame analytics evaluation API to all available metrics.
The motivation is that if the user skips setting an explicit
set of metrics, they get most of the evaluation offering.

iverase committed:
geo_point runtime field implementation (#63164)

Run time field that emits geo points from lat/lon values.

javanna committed:
Mute CoreValuesSourceTypeTests.testDatePrepareRoundingWithDocs (#63970)

Relates to #63969

Commit:
Handle range query edge case (#63397)

Currently when searching with an empty string as lower bound for a range query
on text-based fields we return all documents when 'gte' is used (including the
lower bound) but no documents when 'gt' is used. This might seem
counterintuitive since every value should be greate than the empty string. 
The bug has been fixed in Lucene and this PR adds a test for assuring we observe
the fixed behaviour on searches now.

Closes #63386

javanna committed:
Mute DatafeedJobsIT#testDatafeedTimingStats_DatafeedRecreated (#63974)

Relates to #63973

javanna committed:
Mute FieldSortIT#testCastDate (#63972)

Relates to #63719

pugnascotia committed:
Only subject data streams to allow_auto_create, not auto_create_index

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 20640-auto-create-templates

gwbrown commented on docs/reference/indices/put-component-template.asciidoc:
```diff
@@ -112,6 +112,16 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=settings]
+<<index-creation,`action.auto_create_index`>> cluster setting. If set to
+`true` in a template, then indices can be automatically created using that
+template even if auto-creation of indices is disabled via
+`actions.auto_create_index`. If set to `false`, then indices matching the
```
```suggestion
`actions.auto_create_index`. If set to `false`, then indices or data streams matching the
```

This is the only note I'd make in the documentation, I think.

AthenaEryma reviewed (approved):
This LGTM now, thanks for dealing with all the details around the interaction with data streams! I know multiple rounds of back-and-forth like that can be frustrating.

pugnascotia committed:
Tweak docs

Mention data streams in the docs changes.

Co-authored-by: Gordon Brown <arcsech@gmail.com>

pugnascotia commented:
@jaymode are you happy with these changes now? There's still a "changes requested" on the PR.

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 20640-auto-create-templates

jaymode reviewed (approved):
LGTM

jaymode commented:
> There's still a "changes requested" on the PR.

If you don't mind, can you re-request a review once you feel the PR is ready for me to look at the next time.

martijnvg commented on qa/smoke-test-http/src/test/java/org/elasticsearch/http/AutoCreateIndexIT.java:
```diff
@@ -0,0 +1,126 @@
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hamcrest.Matchers.containsString;
+
+public class AutoCreateIndexIT extends ESRestTestCase {
```
Maybe `distribution/archives/integ-test-zip` is a better place for this test class? Given that it doesn't extend `HttpSmokeTestCase` and this test class doesn't seem to require to start a node with special settings or plugins.

pugnascotia replied:
I'm honestly not sure. I tried before to find out where is the "correct" place for REST tests, but couldn't get a clear answer. I think I'll leave this test alone and start an email thread about it. I feel that it ought to be easier to say where a REST test should live.

martijnvg reviewed (approved):
Left one small comment. LGTM otherwise.

pugnascotia commented:
This ought to be backported after https://github.com/elastic/elasticsearch/pull/63122, I'm trying to find out what is happening with that PR.
